### ISSUE TITLE
Flash damage allocation

### ISSUE THREAD

**Issue by Zero1627** [Aatu/FieryVoid#172]
Flash weapon that hits takes list of possible target systems from section hit, not from hit table. I propose to change that to taking it from hit table. This ges important with presence of unhittable systems AND hittable systems on other sections. 
Also, as MCV outer hit charts do not have structure, current behavior is to strip all systems, and then proceed to PRIMARY section (while  PRIMARY Structure should be allocated, hit chart is not yet depleted).

Tom, any opposition to such change?

**Comment by tksolway** [comment]
Nope, it sounds like that is how it should be done now with hit charts

**Comment by Zero1627** [comment]
done and deployed. 
NOTE: hit chart is not used directly. Instead, list of possible targets is prepared using hit table. Each eligible system has equal chance of being hit.

**Comment by tksolway** [comment]
Is it possible to add items to your generated list multiple times if they occupy multiple places on the hit chart to maintain the probability of hitting that system?  Otherwise, this fix significantly reduces the chance that Flash will hit larger systems like Structure.

**Comment by Zero1627** [comment]
>Otherwise, this fix significantly reduces the chance that Flash will hit larger systems like Structure.
Untrue. Currently there is absolutely no chance (Flash specifically avoided Structure), now the chance exists ;) . I did not modify choice algorithm itself - size of systems played no role in Flash allocation I just added Structure and put systems that exist on hit chart instead of systems techically mounted on section
Choice is made by: 
$roll = Dice::d(sizeof($systems));
...and I did not modify it in any way.

>Is it possible to add items to your generated list multiple times if they occupy multiple places on the hit chart to maintain the probability of hitting that system? 
I thought about that and it's possible. But one would also need to account for multiple instances of a structure (Flash chooses between actual instances, not just system names). If I go for it, doing it properly (eg. using appropriate weights) wouldn't be all that much harder (and would result in much cleaner code as a bonus). But that would require rewriting of significant portion of code.

I MAY do it someday later - but considering that previous approach worked and nobody complained, I went for lowest-cost solution.

Please either close the thread, or comment again indicating that this is important to do (I'll then make it higher priority on my 'to do' list ;) ).

**Comment by Zero1627** [comment]
sorry for torrent of updates, it seems I can't write a clear text at first try...

**Comment by tksolway** [comment]
I think it's pretty important.  Flash weapons are supposed to be swingy.  You get lucky and take off a side, or you don't get so lucky and just strip some weapons.  This way it really is just strip some weapons.  I can try and take a look at it as well, so I don't think it needs to be top Priority, but should be kept open.

**Comment by tksolway** [comment]
I think it's pretty important. Flash weapons are supposed to be swingy. You get lucky and take off a side, or you don't get so lucky and just strip some weapons. This way it really is just strip some weapons. I can try and take a look at it as well, so I don't think it needs to be top Priority, but should be kept open.

**Comment by Zero1627** [comment]
I have already meddled here and have pretty good idea, so will do it, no problem. Just not a top priority. Also, in FV they were specifically implemented to strip systems/weapons, I only changed it with this modification ;)

**Comment by Zero1627** [comment]
Now that You nudged me... I do have a plan of overhaul in mind, and will do it soon - tomorrow if time allows.

**Comment by Zero1627** [comment]
overhauled and simplified, will finish and deploy this week.

**Comment by Zero1627** [comment]
Overhaul continues - this is one big can of worms...
BaseShip is mostly overhauled, but firing procedure needs overhauling too.

I have noted 2 mutually exclusive mechanics:
 - fixing fire from one ship, so all will hit the same section
 - using actual it chance to estimate defensive value of section

I am keeping the first one (despite this being more difficult) as You have indicated it's very important to You in past discussion (note that this was NOT working so far). This means actual hit chance is useless, as it tells nothing about weapons further in line - I'll just use profile, without other mods. 
Using actual hit chance would be great, but only if further shots werre not fixed by the first one.

**Comment by tksolway** [comment]
If you are talking about the estimated hit chance when someone targets a weapon, I think it is fine for it to simply choose one side when it is on an intersection (randomly, or best, not picky).  It is an estimate anyway, and even in tabletop it is the target that gets to choose which side it hits on an intersection for fire from a single hex, so in the real world in these cases the firing player knows there are two options for hit chance based upon the profile of each side.

It is important that the actual fire not be split though.

**Comment by Zero1627** [comment]
>It is important that the actual fire not be split though.

Done! (note that it was not working previously).
EXCEPTION: will only work for direct fire, not ballistics.

>I think it is fine for it to simply choose one side when it is on an intersection (randomly, or best, not picky)

I kept Your algorithm essentially unchanged, except it now uses raw profile instead of modified profile.


Overhaul got way bigger than I expected, but it's finally done. More and more things came out as I was working, asking to be corrected (...as is usual with overhauls ;) ). At last it's done... 

Now, I will need to deploy this monster ;) - hopefully tomorrow, but only if I do have time - there will be LOTS of syntax errors (overhaul was pretty severe, including changing interfaces of some functions - so there's lots of room for it). And then actual bugs which I might have introduced... But - it SHOULD work way better (closer to B5Wars) and tidier. Once debugged at least.

**Comment by tksolway** [comment]
That’s great, Raw profile should be fine, since the changes to it should be the same regardless of which is used.

**Comment by Zero1627** [comment]
Finished!
It ate way more time than it was worth, but now it's done. Some bugs may of course still rear their heads, especially for more exotic weapons...

I streamlined and simplfied procedures of determining hit chance, hit section and hit system, as well as damage dealing itself.
I have also modified procedure of section choice after all.
Notable effects of my changes:
 - all direct fire weapons from one unit will hit same section (except called shots)
 - Flash weapons use actual hit table
 - some specialized routines are incorporated into basic code instead (notably: fighter ballistics no longer need special procedures or hittimg; Matter weapons no longer need special procedures for overkill ($noOverkill attribute added instead)))
 - for section hit choice, profile difference now directly affects toughness (instead of counting ratio toughness/modified profile)
 - same area, base profile is used (not modified by actual shot) because choice is made for many shots, not just this one
